dev/test/bulk-writer.ts

```diff
@@ -549,55 +553,63 @@ describe('BulkWriter', () => {
   describe('500/50/5 support', () => {
     afterEach(() => setTimeoutHandler(setTimeout));
 
-    // TODO(chenbrian): Actually fix this test...
-    it.skip('does not send batches if doing so exceeds the rate limit', async () => {
-      // The test is considered a success if BulkWriter tries to send the second
-      // batch again after a timeout.
+    it('does not send batches if doing so exceeds the rate limit', done => {
+      // The test is considered a success if BulkWriter tries to send the
+      // second batch after a timeout.
 
-      const arrayRange = Array.from(new Array(500), (_, i) => i);
+      const arrayRange = Array.from(new Array(499), (_, i) => i);
       const requests1 = arrayRange.map(i => setOp('doc' + i, 'bar'));
       const responses1 = arrayRange.map(i => successResponse(i));
-      const arrayRange2 = [500, 501, 502, 503, 504];
+      const arrayRange2 = Array.from(new Array(100), (_, i) => i + 500);
       const requests2 = arrayRange2.map(i => setOp('doc' + i, 'bar'));
       const responses2 = arrayRange2.map(i => successResponse(i));
 
-      const bulkWriter = await instantiateInstance([
-        {
-          request: createRequest(requests1),
-          response: mergeResponses(responses1),
-        },
-        {
-          request: createRequest(requests2),
-          response: mergeResponses(responses2),
-        },
-      ]);
-      for (let i = 0; i < 500; i++) {
-        bulkWriter
-          .set(firestore.doc('collectionId/doc' + i), {foo: 'bar'})
-          .then(incrementOpCount);
-      }
-      const flush1 = bulkWriter.flush();
-
-      // Sending this next batch would go over the 500/50/5 capacity, so
-      // check that BulkWriter doesn't send this batch until the first batch
-      // is resolved.
-      let timeoutCalled = false;
-      setTimeoutHandler((_, timeout) => {
-        // Check that BulkWriter has not yet sent the 2nd batch.
-        timeoutCalled = true;
-        expect(requestCounter).to.equal(0);
-        expect(timeout).to.be.greaterThan(0);
+      instantiateInstance(
+        [
+          {
+            request: createRequest(requests1),
+            response: mergeResponses(responses1),
+          },
+          {
+            request: createRequest(requests2),
+            response: mergeResponses(responses2),
+          },
+        ],
+        /* enforceSingleConcurrentRequest= */ true
+      ).then(async bulkWriter => {
+        for (let i = 0; i < 499; i++) {
+          bulkWriter.set(firestore.doc('collectionId/doc' + i), {foo: 'bar'});
+        }
+        bulkWriter.flush();
+
+        // Wait until the initial batch has been sent by BulkWriter. Because
+        // there are floating promises, not waiting here can cause a race
+        // condition where the first batch's request has not been made by the
+        // time the second request is sent.
+        await requestMadeDeferred.promise;
```

Comment:
I am a little confused here... isn't the test meant to send off a lot of writes as fast as possible, and as such it would be beneficial not to wait?

It seems like what we should do is:

```
setTimeoutHandler((_, timeout) => {
  if (!timeoutCalled) {
    timeoutCalled = true;
    // expect(requestCounter).to.equal(0); This is racy without your 
   //  deferred promise, but not needed for the purpose of the test
    expect(timeout).to.be.greaterThan(0);
    // done(); Maybe remove this so we don't have any outstanding 
    //  work when the test completes.
  }
});

for (let i = 0; i < 600; i++) {
   bulkWriter.set(firestore.doc('collectionId/doc' + i), {foo: 'bar'});
}

await bulkWriter.close();
expect(timeoutCalled).to.be.true;

```

Reply:
Post conversation:
The original issue with this simplified test was that it still passes even if the timeout handler is called for the first batch. To fix this, the BulkWriter tests now check that the timeout handler wasn't called after each individual test.

I tried removing the `done()`, but since it's based on the actual clock, the timeout handler will be called multiple times. If we just call the timeout function directly in the overridden handler, we will loop until enough time passes for the rate limiter to allow the request. Though there could still be outstanding work once the test completes, it shouldn't affect other tests since each test creates a new BulkWriter instance.